crates/codegen/parser/runtime/src/query/engine.rs

```diff
@@ -0,0 +1,481 @@
+                        query_number: self.query_number,
+                        bindings,
+                    });
+                }
+                self.query_number += 1;
+            }
+
+            self.advance_to_next_possible_matching_query();
+        }
+
+        None
+    }
+}
+
+trait Combinator {
+    // None -> failed to match, you must backtrack. DO NOT call again
+    // Some(cursor) if cursor.is_complete -> matched, end of input
+    // Some(cursor) if !cursor.is_complete -> matched, more input to go
+    fn next(&mut self) -> Option<Cursor>;
+    fn accumulate_bindings(&self, bindings: &mut HashMap<String, Vec<Cursor>>);
+}
+type CombinatorRef = Box<dyn Combinator>;
+
+struct BindingCombinator {
+    matcher: Rc<BindingMatcher>,
+    cursor: Cursor,
+    child: CombinatorRef,
+}
+
+impl BindingCombinator {
+    fn new(matcher: Rc<BindingMatcher>, cursor: Cursor) -> Self {
+        let child = matcher.child.create_combinator(cursor.clone());
+        Self {
+            matcher,
+            cursor,
+            child,
+        }
+    }
+}
+
+impl Combinator for BindingCombinator {
+    fn next(&mut self) -> Option<Cursor> {
+        self.child.next()
+    }
+
+    fn accumulate_bindings(&self, bindings: &mut HashMap<String, Vec<Cursor>>) {
+        bindings
+            .entry(self.matcher.name.clone())
+            .or_default()
+            .push(self.cursor.clone());
+    }
+}
+
+struct NodeCombinator {
+    matcher: Rc<NodeMatcher>,
+    child: Option<CombinatorRef>,
+    cursor: Cursor,
+    is_initialised: bool,
+}
+
+impl NodeCombinator {
+    fn new(matcher: Rc<NodeMatcher>, cursor: Cursor) -> Self {
+        Self {
+            matcher,
+            child: None,
+            cursor,
+            is_initialised: false,
+        }
+    }
+}
+
+impl Combinator for NodeCombinator {
+    fn next(&mut self) -> Option<Cursor> {
+        if self.cursor.is_completed() {
+            return None;
+        }
+
+        if !self.is_initialised {
+            self.is_initialised = true;
+
+            if !self
+                .cursor
+                .matches_node_selector(&self.matcher.node_selector)
+            {
+                return None;
+            }
+
+            if let Some(child) = self.matcher.child.as_ref() {
+                let mut child_cursor = self.cursor.clone();
+                if !child_cursor.go_to_first_child() {
+                    return None;
+                }
+
+                self.child = Some(child.create_combinator(child_cursor));
+            } else {
+                let mut return_cursor = self.cursor.clone();
+                return_cursor.consume();
+                return Some(return_cursor);
+            }
+        }
+
+        if let Some(child) = self.child.as_mut() {
+            while let Some(cursor) = child.as_mut().next() {
+                if cursor.is_completed() {
+                    let mut return_cursor = self.cursor.clone();
+                    return_cursor.consume();
```

Comment:
not sure I'm following the logic here. For example, why are we returning a cloned/consumed cursor?
Happy to wait for future docs if that answers such questions.

Reply:
This is for symmetry wth the snippet above, but good catch!